### PR TITLE
Add recipe for DOSBox-X

### PR DESCRIPTION
**Pull request by amol-**
DOSBox-X recently upstreamed the changes required to make it compile and work on Haiku

* https://github.com/joncampbell123/dosbox-x/pull/3982
* https://github.com/joncampbell123/dosbox-x/pull/4002

It's not yet perfect, there are some minor graphic glitches and it doesn't yet work correclty with SDL2, but it's usable and I was able to play some games with it. So it makes sense to add a recipe for it given that it's much easier to use than DOSBox itself and it's updated more frequently

### CODE REVIEW

File: games-emulation/dosboxx/additional-files/dosbox.rdef [hunk 1]
```diff
@@ -0,0 +1,116 @@
+
+resource app_flags B_SINGLE_LAUNCH | B_ARGV_ONLY;
+
+resource app_version {
+       major  = 0,
+       middle = 7,
+       minor  = 4,
```
Comment (augiedoggie) [inline games-emulation/dosboxx/additional-files/dosbox.rdef:7]: Where did this version number come from?

File: games-emulation/dosboxx/dosboxx-2022.12.26.recipe [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+	"
+
+BUILD_PREREQUIRES="
+	cmd:aclocal
+	cmd:autoconf
+	cmd:find
+	cmd:gcc$secondaryArchSuffix
+	cmd:ld$secondaryArchSuffix
+	cmd:libtoolize$secondaryArchSuffix
+	cmd:pkg_config
```
Comment (augiedoggie) [inline games-emulation/dosboxx/dosboxx-2022.12.26.recipe:50]:
```suggestion
	cmd:pkg_config$secondaryArchSuffix
```

File: games-emulation/dosboxx/dosboxx-2022.12.26.recipe [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+SOURCE_DIR="dosbox-x-dc350a51928bfede9722b59d40f9ef8d39bf93d0"
+CHECKSUM_SHA256="7cc8ef45c028e91d2ba69fd1885f910db81f9b881776a139c8ba8f7007ecad83"
+ADDITIONAL_FILES="dosbox.rdef"
+
+ARCHITECTURES="all !x86_gcc2"
+SECONDARY_ARCHITECTURES="x86"
+
+PROVIDES="
+	dosboxx$secondaryArchSuffix = $portVersion
+	apps:dosboxx = $portVersion
```
Comment (augiedoggie) [inline games-emulation/dosboxx/dosboxx-2022.12.26.recipe:21]:
```suggestion
	apps:dosbox_x = $portVersion
```

**Review by augiedoggie — CHANGES_REQUESTED** [review]
The folder should be renamed to `dosbox-x` to match [portage layout](http://gpo.zugaina.org/games-emulation/dosbox-x) and the recipe file should be renamed to `dosbox_x-2022.12.26.recipe` with an underscore to replace the dash.

File: games-emulation/dosboxx/additional-files/dosbox.rdef (hunk 1 above)
Comment (amol-) [reply games-emulation/dosboxx/additional-files/dosbox.rdef:7]: Ops, you are correct, this file was copied from DOSBox and I forgot to update it

**Comment by amol-** [comment]
> The folder should be renamed to `dosbox-x` to match [portage layout](http://gpo.zugaina.org/games-emulation/dosbox-x) and the recipe file should be renamed to `dosbox_x-2022.12.26.recipe` with an underscore to replace the dash.

This should now be correct.

File: games-emulation/dosbox-x/dosbox_x-2022.12.26.recipe [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+ARCHITECTURES="all !x86_gcc2"
+SECONDARY_ARCHITECTURES="x86"
+
+PROVIDES="
+	dosbox_x$secondaryArchSuffix = $portVersion
+	apps:dosbox_x = $portVersion
+	"
+
+REQUIRES="
```
Comment (korli) [inline games-emulation/dosbox-x/dosbox_x-2022.12.26.recipe:24]:
```suggestion
REQUIRES="
```

File: games-emulation/dosbox-x/dosbox_x-2022.12.26.recipe [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+BUILD_REQUIRES="
+	haiku${secondaryArchSuffix}_devel
+	devel:libGL$secondaryArchSuffix
+	devel:libglu$secondaryArchSuffix
+	devel:libpng16$secondaryArchSuffix
+	devel:libz$secondaryArchSuffix
+	devel:libncurses$secondaryArchSuffix
+	"
+
+BUILD_PREREQUIRES="
```
Comment (korli) [inline games-emulation/dosbox-x/dosbox_x-2022.12.26.recipe:43]:
```suggestion
BUILD_PREREQUIRES="
```

File: games-emulation/dosbox-x/dosbox_x-2022.12.26.recipe [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+	./build-debug --datarootdir=$dataDir --datadir=$dataDir --bindir=$binDir
+}
+
+INSTALL()
+{
+	make install
+
+	mkdir $appsDir
+	mv $binDir/dosbox-x $appsDir/DOSBox-X
```
Comment (korli) [inline games-emulation/dosbox-x/dosbox_x-2022.12.26.recipe:72]: why not keep the name dosbox-x ?

File: games-emulation/dosbox-x/dosbox_x-2022.12.26.recipe [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+	rm -rf $binDir
+	strip $appsDir/DOSBox-X
+
+	mv /usr/share/dosbox-x $dataDir
+
+	addResourcesToBinaries $portDir/additional-files/dosbox.rdef $appsDir/DOSBox-X
+	addAppDeskbarSymlink $appsDir/DOSBox-X
```
Comment (korli) [inline games-emulation/dosbox-x/dosbox_x-2022.12.26.recipe:79]:
```suggestion
	addAppDeskbarSymlink $appsDir/dosbox-x "DOSBox-X"
```

File: games-emulation/dosbox-x/dosbox_x-2022.12.26.recipe (hunk 6 above)
Comment (amol-) [reply games-emulation/dosbox-x/dosbox_x-2022.12.26.recipe:72]: Consistency with what DOSBox recipe does

File: games-emulation/dosbox-x/additional-files/dosbox.rdef [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+       minor  = 26,
+
+       variety = B_APPV_FINAL,
+       internal = 0,
+
+       short_info = "DOSBox-X",
+       long_info = "A DOS-emulator"
+};
+
+resource app_signature "application/x-vnd.dosbox";
```
Comment (augiedoggie) [inline games-emulation/dosbox-x/additional-files/dosbox.rdef:16]:
```suggestion
resource app_signature "application/x-vnd.dosbox-x";
```

File: games-emulation/dosbox-x/dosbox_x-2022.12.26.recipe [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+BUILD_PREREQUIRES="
+	cmd:aclocal
+	cmd:autoconf
+	cmd:find
+	cmd:gcc$secondaryArchSuffix
+	cmd:ld$secondaryArchSuffix
+	cmd:libtoolize$secondaryArchSuffix
+	cmd:pkg_config$secondaryArchSuffix
+	cmd:make
+	"
```
Comment (augiedoggie) [inline games-emulation/dosbox-x/dosbox_x-2022.12.26.recipe:50]: These entries in the REQUIRES/PROVIDES should be sorted alphabetically(after the haiku*) entry which is always first.

File: games-emulation/dosbox-x/dosbox_x-2022.12.26.recipe (hunk 6 above)
Comment (korli) [reply games-emulation/dosbox-x/dosbox_x-2022.12.26.recipe:70]: sure, this is unfortunate for DOSBox users, who then can't use "dosbox-x" or "dosbox" in a terminal...

File: games-emulation/dosbox-x/dosbox_x-2022.12.26.recipe (hunk 9 above)
Comment (amol-) [reply games-emulation/dosbox-x/dosbox_x-2022.12.26.recipe]: Done, they should now be sorted

**Comment by augiedoggie** [comment]
The package fails to build on 32 bit while trying to compile its own internal copy of SDL1.x.  Perhaps it can be made to use the packaged version of SDL like the regular dosbox recipe does.

```
libtool: compile:  gcc -g -O2 -Iinclude -I/sources/dosbox-x-dc350a51928bfede9722b59d40f9ef8d39bf93d0/vs/sdl/include -D_GNU_SOURCE=1 -Wall -Wno-multichar -c /sources/dosbox-x-dc350a51928bfede9722b59d40f9ef8d39bf93d0/vs/sdl/src/video/bwindow/SDL_sysvideo.cc -o build/SDL_sysvideo.o
In file included from /sources/dosbox-x-dc350a51928bfede9722b59d40f9ef8d39bf93d0/vs/sdl/src/video/bwindow/SDL_sysmouse.cc:27:
/sources/dosbox-x-dc350a51928bfede9722b59d40f9ef8d39bf93d0/vs/sdl/src/video/bwindow/SDL_BWin.h:34:10: fatal error: be/opengl/GLView.h: No such file or directory
   34 | #include <be/opengl/GLView.h>
      |          ^~~~~~~~~~~~~~~~~~~~
compilation terminated.
build-deps:1080: recipe for target 'build/SDL_sysmouse.lo' failed
make: *** [build/SDL_sysmouse.lo] Error 1
make: *** Waiting for unfinished jobs....
In file included from /sources/dosbox-x-dc350a51928bfede9722b59d40f9ef8d39bf93d0/vs/sdl/src/video/bwindow/SDL_sysevents.cc:29:
/sources/dosbox-x-dc350a51928bfede9722b59d40f9ef8d39bf93d0/vs/sdl/src/video/bwindow/SDL_BWin.h:34:10: fatal error: be/opengl/GLView.h: No such file or directory
   34 | #include <be/opengl/GLView.h>
      |          ^~~~~~~~~~~~~~~~~~~~
compilation terminated.
build-deps:1053: recipe for target 'build/SDL_sysevents.lo' failed
make: *** [build/SDL_sysevents.lo] Error 1
In file included from /sources/dosbox-x-dc350a51928bfede9722b59d40f9ef8d39bf93d0/vs/sdl/src/video/bwindow/SDL_sysvideo.cc:28:
/sources/dosbox-x-dc350a51928bfede9722b59d40f9ef8d39bf93d0/vs/sdl/src/video/bwindow/SDL_BWin.h:34:10: fatal error: be/opengl/GLView.h: No such file or directory
   34 | #include <be/opengl/GLView.h>
      |          ^~~~~~~~~~~~~~~~~~~~
compilation terminated.
build-deps:1114: recipe for target 'build/SDL_sysvideo.lo' failed
make: *** [build/SDL_sysvideo.lo] Error 1
Warning: Command '['bash', '-c', '. /wrapper-script']' returned non-zero exit status 1.
Error: Build has failed - stopping.
```

**Comment by amol-** [comment]
> The package fails to build on 32 bit while trying to compile its own internal copy of SDL1.x. Perhaps it can be made to use the packaged version of SDL like the regular dosbox recipe does.
> 
> ```
> libtool: compile:  gcc -g -O2 -Iinclude -I/sources/dosbox-x-dc350a51928bfede9722b59d40f9ef8d39bf93d0/vs/sdl/include -D_GNU_SOURCE=1 -Wall -Wno-multichar -c /sources/dosbox-x-dc350a51928bfede9722b59d40f9ef8d39bf93d0/vs/sdl/src/video/bwindow/SDL_sysvideo.cc -o build/SDL_sysvideo.o
> In file included from /sources/dosbox-x-dc350a51928bfede9722b59d40f9ef8d39bf93d0/vs/sdl/src/video/bwindow/SDL_sysmouse.cc:27:
> /sources/dosbox-x-dc350a51928bfede9722b59d40f9ef8d39bf93d0/vs/sdl/src/video/bwindow/SDL_BWin.h:34:10: fatal error: be/opengl/GLView.h: No such file or directory
>    34 | #include <be/opengl/GLView.h>
>       |          ^~~~~~~~~~~~~~~~~~~~
> compilation terminated.
> build-deps:1080: recipe for target 'build/SDL_sysmouse.lo' failed
> make: *** [build/SDL_sysmouse.lo] Error 1
> make: *** Waiting for unfinished jobs....
> In file included from /sources/dosbox-x-dc350a51928bfede9722b59d40f9ef8d39bf93d0/vs/sdl/src/video/bwindow/SDL_sysevents.cc:29:
> /sources/dosbox-x-dc350a51928bfede9722b59d40f9ef8d39bf93d0/vs/sdl/src/video/bwindow/SDL_BWin.h:34:10: fatal error: be/opengl/GLView.h: No such file or directory
>    34 | #include <be/opengl/GLView.h>
>       |          ^~~~~~~~~~~~~~~~~~~~
> compilation terminated.
> build-deps:1053: recipe for target 'build/SDL_sysevents.lo' failed
> make: *** [build/SDL_sysevents.lo] Error 1
> In file included from /sources/dosbox-x-dc350a51928bfede9722b59d40f9ef8d39bf93d0/vs/sdl/src/video/bwindow/SDL_sysvideo.cc:28:
> /sources/dosbox-x-dc350a51928bfede9722b59d40f9ef8d39bf93d0/vs/sdl/src/video/bwindow/SDL_BWin.h:34:10: fatal error: be/opengl/GLView.h: No such file or directory
>    34 | #include <be/opengl/GLView.h>
>       |          ^~~~~~~~~~~~~~~~~~~~
> compilation terminated.
> build-deps:1114: recipe for target 'build/SDL_sysvideo.lo' failed
> make: *** [build/SDL_sysvideo.lo] Error 1
> Warning: Command '['bash', '-c', '. /wrapper-script']' returned non-zero exit status 1.
> Error: Build has failed - stopping.
> ```

DOSBox-X discourages usage of standard SDL1 as their fork includes specific patches.

I'm working on fixing some issues when using SDL2, which will allow to reuse the system one as DOSBox-X officially supports usage of system SDL2, only for SDL1 they don't.

Meanwhile I would simply disable support for 32bit systems. I'll re-enable it as soon as I get fixes for SDL2 done.

**Comment by amol-** [comment]
> 
> Meanwhile I would simply disable support for 32bit systems. I'll re-enable it as soon as I get fixes for SDL2 done.

Also that seems some missing dependency more than a problem with 32bit systems 🤔

**Comment by amol-** [comment]
Any feedback on the idea of disabling for 32bit systems and releasing the recipe?

**Comment by augiedoggie** [comment]
> Any feedback on the idea of disabling for 32bit systems and releasing the recipe?

It's fine by me.  It sounded like you were going to make the change and so I was waiting for an update to the pull request.

**Comment by amol-** [comment]
> > Any feedback on the idea of disabling for 32bit systems and releasing the recipe?
> 
> It's fine by me. It sounded like you were going to make the change and so I was waiting for an update to the pull request.

ah, no, I meant I'm in the process of fixing haiku related issues when building DOSBox-X with sdl2. But that's bound to how much time I have to invest into that. So it would make sense to release the recipe meanwhile. I can cut a new recipe update once SDL2 issues are fixed and we can build DOSBox-X with SDL2

**Pull request merged**